transformer_lens/components.py

```diff
@@ -1016,25 +1017,41 @@ def forward(
         """
         resid_pre = self.hook_resid_pre(resid_pre)  # [batch, pos, d_model]
 
-        query_input = resid_pre
-        key_input = resid_pre
-        value_input = resid_pre
+        def add_head_dimension(tensor):
+            return einops.repeat(
+                tensor,
+                "batch pos d_model -> batch pos n_heads d_model",
+                n_heads=self.cfg.n_heads,
+            ).clone()
+
+        attn_in = (
+            resid_pre
+            if not self.cfg.use_attn_in
+            else self.hook_attn_in(add_head_dimension(resid_pre.clone()))
+        )
 
         if self.cfg.use_split_qkv_input:
-
-            def add_head_dimension(tensor):
-                return einops.repeat(
-                    tensor,
-                    "batch pos d_model -> batch pos n_heads d_model",
-                    n_heads=self.cfg.n_heads,
-                ).clone()
-
-            query_input = self.hook_q_input(add_head_dimension(query_input))
-            key_input = self.hook_k_input(add_head_dimension(key_input))
-            value_input = self.hook_v_input(add_head_dimension(value_input))
-
+            query_input = self.hook_q_input(
+                add_head_dimension(attn_in)
+                if not self.cfg.use_attn_in
+                else attn_in  # we've already added the extra dimension if we're using attn_in!
+            )
+            key_input = self.hook_k_input(
+                add_head_dimension(attn_in)
+                if not self.cfg.use_attn_in
+                else attn_in  # we've already added the extra dimension if we're using attn_in!
+            )
+            value_input = self.hook_v_input(
+                add_head_dimension(attn_in)
+                if not self.cfg.use_attn_in
+                else attn_in  # we've already added the extra dimension if we're using attn_in!
+            )
```

Comment:
This should be equivalent, unless I'm missing something?
```suggestion
            # we've already added the extra dimension if we're using attn_in!
            if not self.cfg.use_attn_in:
                attn_in = add_head_dimension(attn_in)
            query_input = self.hook_q_input(attn_in)
            key_input = self.hook_k_input(attn_in)
            value_input = self.hook_v_input(attn_in)
```

Reply:
No, this specific idea shouldn't be used because inplace operations on `attn_in` will cause problems, similar to [here](https://github.com/neelnanda-io/TransformerLens/issues/157). In short, with your change, naively zero ablating `query_input` would zero ablate `key_input` and `value_input` too.

However, thanks for writing as I just noticed that I should be cloning `attn_in` here for the same reason!

Reply:
That makes sense, thanks. My issue was the duplicated code for q/k/v_input so I'm happy now.
Personally, I'd have `add_head_dimension` without cloning and clone explicitly. Also, should we make the same changes to BertBlock?